lib/internal/async_local_storage/native.js

```diff
@@ -0,0 +1,47 @@
+'use strict';
+
+const {
+  ReflectApply,
+} = primordials;
+
+const AsyncContextFrame = require('internal/async_context_frame');
+const { AsyncResource } = require('async_hooks');
+
+class AsyncLocalStorage {
+  static bind(fn) {
+    return AsyncResource.bind(fn);
+  }
+
+  static snapshot() {
+    return AsyncLocalStorage.bind((cb, ...args) => cb(...args));
+  }
+
+  disable() {
+    AsyncContextFrame.disable(this);
```

Comment:
This `disable` variant seems quite different to the other one as this version effects only the current frame (which might be even `undefined`) whereas the other variant disables the ALS instance globally

Reply:
That's not quite accurate as `run`/`enterWith` will enable it again automatically. The existing `disable` also doesn't actually change the value stored on the associated resource so, for example, you could call `disable()` and then follow it with a `run()` and the previously entered context _before_ the disable would be restored after the `run()` exits, which is kind of a bug. I'm intentionally leaving the _existing_ ALS behaviour as-is, despite having several bugs. Those incorrect behaviours can be fixed in follow-up PRs.

The API is also considered experimental, so _can_ be changed at any time. Support for the existing experimental APIs is somewhat best-effort and I would _prefer_ to deprecate these APIs entirely unless some specific effort goes into fixing and stabilizing these APIs.

I don't think exact bug-for-bug reproduction of the existing ALS should be a goal here. The stabilized parts should be expected to behave the same, but the rest is best effort and likely to deviate a bit as the new model allows fixing some issues in the design of the prior model.

Reply:
Yes, `run/enterWith` reenable it but e.g. a timers or task_queue items do not use these APIs.

Anyhow, I agree that `disable`/`exit` APIs are strange (e.g. `exit` isn't the pendent of `enter`,...) already now. I also think they have no real use case besides testing maybe. So fine for me if we have behavior changes or even remove them later on.

Just want to make sure that these differences are not slipping in unintended and people complain. Maybe worth to mention this in the initial PR comment and/or commit message to avoid people say this was a silent change.

Reply:
I intend to do some follow-up improvements, like there's currently a bunch of enabled checks in hot-paths that I want to restructure somewhat to have already made that check ahead of time.

I _do_ think we should actually _keep_ the `exit` behaviour at least, but I want to work on making the behaviour more coherent and consistent.

I'm not quite so convinced we should be keeping `disable` as it not being scoped doesn't match up evenly with the rest of the API and so the behaviour is a bit confusing given that it gets re-enabled automatically in certain cases. If it required an _explicit_ enable then it might be more reasonable, but with the _implicit_ re-enable it functions essentially the same as `exit` but in a much more edge-case-y way.

Reply:
`exit()` without the internal `disable()` is mostly an `enter(undefined)`.
I would prefer a single `set()` instead but I guess it's a bit too late :o)